mbf_costmap_nav/src/mbf_costmap_nav/free_pose_search.cpp

```diff
@@ -0,0 +1,298 @@
+    for (const auto& theta : thetas)
+    {
+      sol.pose.theta = theta;
+      SearchState search_state = getFootprintState(costmap_2d, footprint, sol.pose);
+
+      switch (search_state.state)
+      {
+        case SearchState::FREE:
+        case SearchState::INSCRIBED:
+          if (viz)
+          {
+            viz->addSolution(sol.pose, footprint);
+          }
+          return { sol.pose, search_state };
+          break;
+        case SearchState::UNKNOWN:
+        case SearchState::OUTSIDE:
+          if (!found_unknown)
+          {
+            // we save the first unknown/outside pose, but we continue searching for a free pose
+            sol = { sol.pose, search_state };
+            found_unknown = true;
+          }
+          break;
+        case SearchState::LETHAL:
+          // if we didn't find a free pose or unknown/outside, we return lethal
+          if (!found_unknown)
+          {
+            sol.search_state = search_state;
+          }
+          if (viz)
+          {
+            viz->addBlocked(sol.pose, footprint);
+          }
+          break;
+        default:
+          ROS_ERROR_STREAM("Unknown state: " << search_state.state);
+          break;
+      }
+    }
+  }
+
+  ROS_DEBUG_STREAM_COND_NAMED(sol.search_state.state == SearchState::LETHAL, LOGNAME.data(),
+                              "No valid orientation found for pose x-y-theta ("
+                                  << pose_2d.x << ", " << pose_2d.y << ", " << pose_2d.theta << ") with tolerance "
+                                  << config.angle_tolerance << " and increment " << config.angle_increment);
+
+  ROS_DEBUG_STREAM_COND_NAMED(sol.search_state.state == SearchState::UNKNOWN, LOGNAME.data(),
+                              "Solution is in unknown space for pose x-y-theta ("
+                                  << pose_2d.x << ", " << pose_2d.y << ", " << pose_2d.theta << ") with tolerance "
+                                  << config.angle_tolerance << " and increment " << config.angle_increment);
+  return sol;
+}
+
+SearchSolution FreePoseSearch::search() const
+{
+  const auto costmap2d = costmap_.getCostmap();
+
+  // lock costmap so content doesn't change while adding cell costs
+  boost::unique_lock<costmap_2d::Costmap2D::mutex_t> lock(*(costmap2d->getMutex()));
+
+  std::unordered_set<int> in_queue_or_visited;
+
+  const std::vector<geometry_msgs::Point> footprint =
+      safetyPadding(costmap_, config_.use_padded_fp, config_.safety_dist);
+
+  SearchSolution sol;
+  sol.pose.theta = config_.goal.theta;
+  bool found_unknown{ false };
+
+  // initializing queue with the goal cell
+  Cell start;
+  if (costmap2d->worldToMap(config_.goal.x, config_.goal.y, start.x, start.y))
+    start.cost = costmap2d->getCost(start.x, start.y);
+  else
+    start.cost = costmap_2d::FREE_SPACE;  // outside of the map; TODO: if set to NO_INFORMATION, we won't find a solution
```

Comment:
S: would be nicer with NO_INFORMATION, but then we need to change the search code